### PR TITLE
Mark waiting upload parts as cancelled when canceling a multipart upload

### PR DESCRIPTION
**Pull request by colinhumber**
An NSURLSession that has been marked as invalidated via `finishTasksAndInvalidate` will only fully invalidate when all tasks associated with it have been completed. Currently, only in-progress parts
are being cancelled and thus marked as completed. Any waiting tasks remain in the `NSURLSessionTaskStateSuspended` state. As per the documentation, the session will only be flagged as invalidated, calling
the `[NSURLSession session:didBecomeInvalidWithError:]` when all tasks have a state of `NSURLSessionTaskStateCompleted`.

### CODE REVIEW

**Comment by colinhumber** [comment]
https://github.com/aws-amplify/aws-sdk-ios/issues/2280

**Comment by palpatim** [comment]
Also, could you please change the base branch of this PR to `develop` rather than `master`? Thanks!

**Comment by colinhumber** [comment]
@palpatim Base branch changed! I was looking at supplementing the `testCancelMultipartUpload` test in `AWSTransferUtilityTests.swift` to double check that the `AWSS3TransferUtilityURLSessionDidBecomeInvalid` notification is posted, but I get a bunch of compile errors when running the tests. The `credentials.json` file is missing and I'm not sure what it's supposed to look like. Removing the reference to it adds a number of additional linking errors. I'm at a bit of a loss.

**Comment by kneekey23** [comment]
hi @colinhumber sorry for the delay in response and apologies for our lack of a contributing guide with how to set this up (I plan to fix this in the near future to make it easier to contribute). We do however very much appreciate your contribution! So now to get to your question, yes `credentials.json` is a file that is apart of our codebase that we do not check in (cuz security). To recreate the same file in your repo you want to create it inside AWSCoreTests> Resources>credentials.json and it should have the following format:
```
{
    "accessKey": "XXXXXXXXXXXXXX",
    "secretKey": "XXXXXXXXXXXXXXXXXXXXXXXXX",
    "accountId": "XXXXXXXXXXXX",
    "identityPoolId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "unauthIdentityPoolId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "authRoleArn": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX", 
    "unauthRoleArn":"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "amazon-connect-instance-id": "XXXXXXXXXXXXXXXXXXXXX",
    "amazon-connect-destination-dial-in-phone-number": "XXXXXXXXXX",
    "anotherAccountId": "XXXXXXXXXXXXX",
    "apiEndpoint": "XXXXXXXXXXXXXXXXXXXXXXXXX",
    "facebookAppId": "XXXXXXXXXXXX",
    "facebookAppSecret": "XXXXXXXXXXXXXXXXXXXX",
    "iot-custom-authorizer-name": "XXXXXXXXXX",
    "iot-custom-authorizer-token-key-name": "XXXXXXXXXXXX",
    "iot-custom-authorizer-token-signature": "XXXXXXXXXXXXXXXXXXXXXXXX",
    "iot-custom-authorizer-token-value": "allow",
    "iot-us-east1-endpoint": "XXXXXXXXXXXXXXXXXXXXXX",
    "iot-us-east2-endpoint": "XXXXXXXXXXXXXXXXXXXXXX",
    "pinpointAppId": "XXXXXXXXXXXXXXXXXXX",
    "sageMaker-endpoint": "XXXXXXXXXXXXXXXX",
    "snsTestAppArn": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "textract-s3-bucket-name": "XXXXXXXXXXXXXXXXXXXXX",
    "textract-s3-object-name": "XXXXXXXXXXXXXXXXXXXXXX",
    "verifiedEmailAddress": "XXXXXXXXXXXXXXX"
}
```
This is the full file  ( in case you would like to make future contributions later on ;) ) which would be quite a bit of AWS resource creation. However, for the purpose of just adding to the transfer utility tests you just need to fill out the below keys and you should be able to run the transfer utility tests just fine (tested it myself):
```
    "accessKey": "XXXXXXXXXXXXXX",
    "secretKey": "XXXXXXXXXXXXXXXXXXXXXXXXX",
    "accountId": "XXXXXXXXXXXX",
    "identityPoolId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "unauthIdentityPoolId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "authRoleArn": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX", 
    "unauthRoleArn":"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
```
The tests themselves create and delete the bucket after the tests so you just need an IAM role with S3 access. Please let me know if you need further assistance with setting up to run these tests and I would be happy to help, again thanks for your contribution! :)

**Comment by colinhumber** [comment]
@kneekey23 Awesome! Thanks for the info :D I'll get some test written up and running.

**Comment by kneekey23** [comment]
Sounds great, let me know if you run into any issues!

**Comment by colinhumber** [comment]
@kneekey23 I was able to get my user setup but I'm running into issues with the test setup where the bucket is being created. In `AWSS3TransferUtilityTests.swift` on line 94, specifically. I get a couple of log messages looking like it's trying to create a bucket using a generic set of permissions, not the user I setup to be used in `credentials.json`.

```
2020-02-24 14:56:40:607 AWSAllTestsHost[39933:7795617] Couldn't locate the awsconfiguration.json file. Skipping load of awsconfiguration.json.

2020-02-24 14:56:41:424 AWSAllTestsHost[39933:7795846] Response body:
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>1EF87714C5CA0166</RequestId><HostId>aXlVoXLqIJD40N2z6dsuan50NQGGO8wn+MITW7dd3jMFWlnc+vNfSRLl5FbJQxaxquhdmmDiFGc=</HostId></Error>
```

**Comment by kneekey23** [comment]
hi @colinhumber did you make sure you changed the bucketname? Forgot to mention that. Buckets in S3 are globally unique so you need to make sure you have a different bucket name first and foremost. Then if it's still failing, can you paste your iam role's policy so I can see it and make sure it has the right access?

**Comment by colinhumber** [comment]
@kneekey23 Yeah, we tried changing the bucket names to no avail.

I have my IAM user with `AmazonS3FullAccess` permissions. Here's the policy JSON:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "s3:*",
            "Resource": "*"
        }
    ]
}
```

Looking at the code that creates the bucket in the `AWSS3TestHelper`, it appears to be using the default S3 client instead of the identity providers I setup. I might just be looking in the wrong area, though 😬

**Comment by colinhumber** [comment]
Oh nevermind. I just found where the credential provider is set internally for the `defaultS3` service.

**Comment by colinhumber** [comment]
One thing I just remembered we were a little confused on was where to find the `unauthIdentityPoolId`. I couldn't find it in the admin panel anywhere, just the `identityPoolId`.

**Comment by kneekey23** [comment]
@colinhumber we have 2 identity pools (one is says auth and the other says unauth but they are set up with the same roles and same config (no clue why, don't ask lol), just paste the same identity pool id for the unauth ones.

**Comment by kneekey23** [comment]
looks like the code is using the keys from credential.json and specifically the identityPoolId with the arns of the roles. 
```if (![AWSServiceManager defaultServiceManager].defaultServiceConfiguration) {
        NSString *filePath = [[NSBundle bundleForClass:[self class]] pathForResource:@"credentials"
                                                                              ofType:@"json"];
        NSDictionary *credentialsJson = [NSJSONSerialization JSONObjectWithData:[NSData dataWithContentsOfFile:filePath]
                                                                        options:NSJSONReadingMutableContainers
                                                                          error:nil];

        AWSCognitoCredentialsProvider *credentialsProvider = [[AWSCognitoCredentialsProvider alloc] initWithRegionType:AWSRegionUSEast1
                                                                                                        identityPoolId:credentialsJson[@"identityPoolId"]
                                                                                                         unauthRoleArn:credentialsJson[@"unauthRoleArn"]
                                                                                                           authRoleArn:credentialsJson[@"authRoleArn"]
                                                                                               identityProviderManager:nil];

        AWSServiceConfiguration *configuration = [[AWSServiceConfiguration alloc] initWithRegion:AWSRegionUSEast1
                                                                             credentialsProvider:credentialsProvider];
        [AWSServiceManager defaultServiceManager].defaultServiceConfiguration = configuration;
    }
```

Our polices for our roles are below (with account number obscured):
Auth role has 2 policies:
**Policy 1**:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "mobileanalytics:PutEvents",
                "cognito-sync:*",
                "cognito-identity:*"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```
**Policy 2**:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": "cognito-sync:QueryRecords",
            "Resource": "*"
        },
        {
            "Sid": "VisualEditor1",
            "Effect": "Allow",
            "Action": "cognito-sync:*",
            "Resource": "arn:aws:cognito-sync:*:*:identitypool/*/identity/*"
        },
        {
            "Sid": "VisualEditor2",
            "Effect": "Allow",
            "Action": "cognito-sync:*",
            "Resource": "arn:aws:cognito-sync:*:*:identitypool/*"
        },
        {
            "Sid": "VisualEditor3",
            "Effect": "Allow",
            "Action": "cognito-sync:*",
            "Resource": "arn:aws:cognito-sync:*:*:identitypool/*/identity/*/dataset/*"
        }
    ]
}
```
Clearly the Auth role isn't being used for the bucket creation.
Looks like the Unauth role has 2 policies and the first one **is** being used for the integ tests
**Policy 1**:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "apigateway:GET",
                "autoscaling:AttachInstances",
                "cloudwatch:DescribeAlarmHistory",
                "ec2:AttachVolume",
                "elasticloadbalancing:ConfigureHealthCheck",
                "execute-api:Invoke",
                "firehose:PutRecord",
                "firehose:PutRecordBatch",
                "kinesis:CreateStream",
                "kinesis:DescribeStream",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:PutRecord",
                "kinesis:PutRecords",
                "kinesisvideo:CreateStream",
                "kinesisvideo:DeleteStream",
                "kinesisvideo:GetDataEndpoint",
                "kinesisvideo:GetHLSStreamingSessionURL",
                "kinesisvideo:GetMediaForFragmentList",
                "kms:CancelKeyDeletion",
                "kms:CreateAlias",
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:DisableKeyRotation",
                "kms:Encrypt",
                "kms:ScheduleKeyDeletion",
                "lambda:GetFunction",
                "lambda:InvokeFunction",
                "machinelearning:GetMLModel",
                "machinelearning:Predict",
                "rekognition:CreateCollection",
                "rekognition:IndexFaces",
                "s3:GetObject",
                "sagemaker:InvokeEndpoint",
                "sdb:BatchPutAttributes",
                "sdb:CreateDomain",
                "sdb:DeleteDomain",
                "sdb:PutAttributes",
                "sdb:Select",
                "ses:SendEmail",
                "sqs:GetQueueAttributes"
            ],
            "Resource": [
                "arn:aws:apigateway:us-east-1:XXXXXXXXXXX:*",
                "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
                "arn:aws:cloudwatch:*:*:alarm:*",
                "arn:aws:ec2:*:*:instance/*",
                "arn:aws:ec2:*:*:volume/*",
                "arn:aws:elasticloadbalancing:*:*:loadbalancer/*",
                "arn:aws:execute-api:us-east-1:XXXXXXXXX:*",
                "arn:aws:firehose:us-east-1:XXXXXXXXXXX:deliverystream/test-permanent-firehose",
                "arn:aws:kinesis:us-east-1:XXXXXXXXXXXX:stream/*",
                "arn:aws:kinesisvideo:us-east-1:XXXXXXXXXXX:stream/*/*",
                "arn:aws:kms:us-east-1:XXXXXXXXXXX:alias/*",
                "arn:aws:kms:us-east-1:XXXXXXXXXXX:key/*",
                "arn:aws:lambda:us-east-1:XXXXXXXXXXX:function:*",
                "arn:aws:machinelearning:us-east-1:XXXXXXXXXXX:mlmodel/ml-UJdLGgO6l44",
                "arn:aws:rekognition:us-east-1:XXXXXXXXXXX:collection/*",
                "arn:aws:s3:::aws-sdk-ios-test-rekognition/family_thumb.jpg",
                "arn:aws:sagemaker:us-east-1:XXXXXXXXXXX:endpoint/xgboost-2019-06-11-20-15-13-355",
                "arn:aws:sdb:us-east-1:XXXXXXXXXXX:domain/*",
                "arn:aws:ses:us-east-1:3XXXXXXXXXXX:identity/XXXXXXXXXXX",
                "arn:aws:sqs:us-east-1:XXXXXXXXXXX:*"
            ]
        },
        {
            "Sid": "IoTStreamTest",
            "Effect": "Allow",
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:GetObjectVersion",
                "iot:CreateStream",
                "iot:DeleteStream",
                "iot:DescribeStream"
            ],
            "Resource": [
                "arn:aws:s3:::ios-sdk-iot-ota-integ-test/*",
                "arn:aws:s3:::ios-sdk-iot-ota-integ-test*",
                "arn:aws:iot:*:*:stream/*"
            ]
        },
        {
            "Sid": "VisualEditor1",
            "Effect": "Allow",
            "Action": "ses:SendEmail",
            "Resource": "arn:aws:ses:us-east-1:XXXXXXXXXXX:identity/XXXXXXXXXXX:"
        },
        {
            "Sid": "VisualEditor2",
            "Effect": "Allow",
            "Action": [
                "autoscaling:DescribeAccountLimits",
                "cloudwatch:GetMetricStatistics",
                "cloudwatch:ListMetrics",
                "comprehend:BatchDetectEntities",
                "comprehend:DetectDominantLanguage",
                "comprehend:DetectEntities",
                "comprehend:DetectKeyPhrases",
                "comprehend:DetectSentiment",
                "connect:*",
                "dynamodb:*",
                "ec2:DescribeImages",
                "ec2:DescribeInstances",
                "elasticloadbalancing:DescribeLoadBalancers",
                "firehose:ListDeliveryStreams",
                "iot:*",
                "kinesis:ListStreams",
                "kinesisvideo:ListFragments",
                "kinesisvideo:ListStreams",
                "kms:CreateKey",
                "lambda:ListFunctions",
                "mobiletargeting:PutEvents",
                "mobiletargeting:UpdateEndpoint",
                "polly:DeleteLexicon",
                "polly:GetSpeechSynthesisTask",
                "polly:PutLexicon",
                "polly:StartSpeechSynthesisTask",
                "polly:SynthesizeSpeech",
                "s3:*",
                "sdb:ListDomains",
                "ses:GetSendQuota",
                "ses:VerifyEmailIdentity",
                "sns:GetEndpointAttributes",
                "sns:ListTopics",
                "sqs:ListQueues",
                "textract:AnalyzeDocument",
                "transcribe:*",
                "translate:TranslateText"
            ],
            "Resource": "*"
        },
        {
            "Sid": "VisualEditor3",
            "Effect": "Allow",
            "Action": "elasticloadbalancing:ConfigureHealthCheck",
            "Resource": "arn:aws:elasticloadbalancing:*:*:loadbalancer/*"
        },
        {
            "Sid": "IoTStreamingTest",
            "Effect": "Allow",
            "Action": "iam:PassRole",
            "Resource": "arn:aws:iam::XXXXXXXXXXX:role/OTAIntegTest"
        }
    ]
}
```
**Policy 2**:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "mobileanalytics:PutEvents",
                "cognito-sync:*"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

Make sure that your Unauth role has full S3 access. My presumption here is that your Auth role has the access and looks like from these policies, the Unauth one is doing the work.

**Comment by colinhumber** [comment]
@kneekey23 Awesome, I'll give those a check and let you know!

**Comment by colinhumber** [comment]
@kneekey23 That did it! Thanks so much for all the help.

**Comment by colinhumber** [comment]
@palpatim Added a new test to track session invalidation when removing a transfer utility.

File: AWSS3Tests/AWSS3TransferUtilityTests.swift [hunk 1]
```diff
@@ -1600,6 +1600,60 @@ class AWSS3TransferUtilityTests: XCTestCase {
...
+
+        sleep(2)
+
+        let sessionInvalidationExpectation = XCTNSNotificationExpectation(name: .AWSS3TransferUtilityURLSessionDidBecomeInvalid, object: nil, notificationCenter: NotificationCenter.default)
+
+        // remove the utility and wait for the session invalidation notification
+        AWSS3TransferUtility.remove(forKey: "to-remove")
+
+        wait(for: [sessionInvalidationExpectation], timeout: 10)
+    }
```
[the remark this answers is not in the document]
Comment (colinhumber) [reply AWSS3Tests/AWSS3TransferUtilityTests.swift]: Yeah, before the change the notification was never firing as the session wasn't being invalidated due to the waiting tasks never being cancelled.

[the remark this answers is not in the document]
Comment (palpatim) [reply AWSS3Tests/AWSS3TransferUtilityTests.swift]: Got it, thanks! We'll merge this in and get it into the next release.

[the remark this answers is not in the document]
Comment (colinhumber) [reply AWSS3Tests/AWSS3TransferUtilityTests.swift]: Awesome! Thanks for all the help!

**Pull request merged**